packages/mobile/src/exchange/saga.ts

```diff
@@ -1,5 +1,248 @@
+    if (minimumTakerAmount.isGreaterThan(updatedTakerAmount)) {
+      Logger.error(
+        TAG,
+        `Not receiving enough ${makerToken} due to change in exchange rate. Exchange failed.`
+      )
+      yield put(showError(ErrorMessages.EXCHANGE_RATE_CHANGE))
+      return
+    }
+
+    const takerTokenContract =
+      makerToken === CURRENCY_ENUM.DOLLAR ? goldTokenContract : stableTokenContract
+    const convertedTakerAmount: BigNumber = roundDown(
+      yield call(convertToContractDecimals, minimumTakerAmount, takerTokenContract),
+      0
+    )
+    Logger.debug(
+      TAG,
+      `Will receive at least ${convertedTakerAmount} 
+      wei for ${convertedMakerAmount} wei of ${makerToken}`
+    )
+
+    let approveTx
+    if (makerToken === CURRENCY_ENUM.GOLD) {
+      approveTx = goldTokenContract.methods.approve(
+        exchangeContract._address,
+        convertedMakerAmount.toString()
+      )
+    } else if (makerToken === CURRENCY_ENUM.DOLLAR) {
+      approveTx = stableTokenContract.methods.approve(
+        exchangeContract._address,
+        convertedMakerAmount.toString()
+      )
+    } else {
+      Logger.error(TAG, `Unexpected maker token ${makerToken}`)
+      return
+    }
+    yield call(sendTransaction, approveTx, account, TAG, 'approval')
+    Logger.debug(TAG, `Transaction approved: ${util.inspect(approveTx.arguments)}`)
+
+    const tx = exchangeContract.methods.exchange(
+      convertedMakerAmount.toString(),
+      convertedTakerAmount.toString(),
+      sellGold
+    )
+
+    if (!txId) {
+      Logger.error(TAG, 'No txId. Did not exchange.')
+      return
+    }
+    yield call(sendAndMonitorTransaction, txId, tx, account)
+  } catch (error) {
+    Logger.error(TAG, 'Error doing exchange', error)
+    if (txId) {
+      yield put(removeStandbyTransaction(txId))
+    }
+
+    if (error.message === ErrorMessages.INCORRECT_PIN) {
+      yield put(showError(ErrorMessages.INCORRECT_PIN))
+    } else {
+      yield put(showError(ErrorMessages.EXCHANGE_FAILED))
+    }
+  }
+}
+
+function* createStandbyTx(
+  makerToken: CURRENCY_ENUM,
+  makerAmount: BigNumber,
+  exchangeRate: BigNumber,
+  account: string
+) {
+  const takerAmount = getTakerAmount(makerAmount, exchangeRate, 2)
+  const txId = generateStandbyTransactionId(account)
+  yield put(
+    addStandbyTransaction({
+      id: txId,
+      type: TransactionTypes.EXCHANGE,
+      status: TransactionStatus.Pending,
+      inSymbol: makerToken,
+      inValue: makerAmount.toString(),
+      outSymbol: makerToken === CURRENCY_ENUM.DOLLAR ? CURRENCY_ENUM.GOLD : CURRENCY_ENUM.DOLLAR,
+      outValue: takerAmount.toString(),
+      timestamp: Math.floor(Date.now() / 1000),
+    })
+  )
+  return txId
+}
+
+async function convertToContractDecimals(value: BigNumber | string | number, contract: any) {
+  // TODO(Rossy): Move this function to SDK and cache this decimals amount
+  const decimals = await contract.methods.decimals().call()
```

Comment:
Can this use contract kit too?

Reply:
This function is only used for the exchange transaction (separate from viewing exchange rate), which will I plan to migrate to contractkit in a future PR, I will update this to use contractkit then. 
In the mean time changing this function would break it, as the contract passed in will be a web3 contract, not a contractkit contract.

Reply:
Updated!